src/algebraic_geometry/AffineScheme.lean

```diff
@@ -307,6 +307,125 @@ begin
+  ∃ f : X.presheaf.obj (op U), X.basic_open f ⊆ V ∧ ↑x ∈ X.basic_open f :=
+begin
+  haveI : is_affine _ := hU,
+  obtain ⟨_, ⟨_, ⟨r, rfl⟩, rfl⟩, h₁, h₂⟩ := (is_basis_basic_open (X.restrict U.open_embedding))
+    .exists_subset_of_mem_open _ ((opens.map U.inclusion).obj V).prop,
+  swap, exact ⟨x, h⟩,
+  have : U.open_embedding.is_open_map.functor.obj ((X.restrict U.open_embedding).basic_open r)
+    = X.basic_open (X.presheaf.map (eq_to_hom U.open_embedding_obj_top.symm).op r),
+  { refine (is_open_immersion.image_basic_open (X.of_restrict U.open_embedding) r).trans _,
+    erw ← Scheme.basic_open_res_eq _ _ (eq_to_hom U.open_embedding_obj_top).op,
+    rw [← comp_apply, ← category_theory.functor.map_comp, ← op_comp, eq_to_hom_trans,
+      eq_to_hom_refl, op_id, category_theory.functor.map_id],
+    erw PresheafedSpace.is_open_immersion.of_restrict_inv_app,
+    congr },
+  use X.presheaf.map (eq_to_hom U.open_embedding_obj_top.symm).op r,
+  rw ← this,
+  exact ⟨set.image_subset_iff.mpr h₂, set.mem_image_of_mem _ h₁⟩,
+  exact x.prop,
+end
+
+instance {X : Scheme} {U : opens X.carrier} (f : X.presheaf.obj (op U)) :
+  algebra (X.presheaf.obj (op U)) (X.presheaf.obj (op $ X.basic_open f)) :=
+(X.presheaf.map (hom_of_le $ RingedSpace.basic_open_subset _ f : _ ⟶ U).op).to_algebra
+
+lemma PresheafedSpace.is_open_immersion.is_iso_of_subset {C : Type*} [category C]
+  {X Y : PresheafedSpace C} (f : X ⟶ Y) [H : PresheafedSpace.is_open_immersion f]
+  (U : opens Y.carrier) (hU : (U : set Y.carrier) ⊆ set.range f.base) : is_iso (f.c.app $ op U) :=
+begin
+  have : U = H.base_open.is_open_map.functor.obj ((opens.map f.base).obj U),
+  { ext1,
+    exact (set.inter_eq_left_iff_subset.mpr hU).symm.trans set.image_preimage_eq_inter_range.symm },
+  convert PresheafedSpace.is_open_immersion.c_iso ((opens.map f.base).obj U),
+end
+
+lemma is_affine_open.from_Spec_preimage_basic_open {X : Scheme} {U : opens X.carrier}
+  (hU : is_affine_open U) (f : X.presheaf.obj (op U)) :
+  (opens.map hU.from_Spec.val.base).obj (X.basic_open f) =
+    RingedSpace.basic_open _ (Spec_Γ_identity.inv.app (X.presheaf.obj $ op U) f) :=
+begin
+  erw LocallyRingedSpace.preimage_basic_open,
+  refine eq.trans _ (RingedSpace.basic_open_res_eq (Scheme.Spec.obj $ op $ X.presheaf.obj (op U))
+    .to_LocallyRingedSpace.to_RingedSpace (eq_to_hom hU.from_Spec_base_preimage).op _),
+  congr,
+  rw ← comp_apply,
+  congr,
+  erw ← hU.Spec_Γ_identity_hom_app_from_Spec,
+  rw iso.inv_hom_id_app_assoc,
+end
+
+/-- The canonical map `Γ(𝒪ₓ, D(f)) ⟶ Γ(Spec 𝒪ₓ(U), D(Spec_Γ_identity.inv f))`
+This is an isomorphism, as witnessed by an `is_iso` instance. -/
+def basic_open_sections_to_affine {X : Scheme} {U : opens X.carrier} (hU : is_affine_open U)
+  (f : X.presheaf.obj (op U)) : X.presheaf.obj (op $ X.basic_open f) ⟶
+    (Scheme.Spec.obj $ op $ X.presheaf.obj (op U)).presheaf.obj
+      (op $ Scheme.basic_open _ $ Spec_Γ_identity.inv.app (X.presheaf.obj (op U)) f) :=
```

Comment:
This is quite a long description. Can we get closer to the way the docstring looks, using notation or definitions?
Can we make `𝒪 X` notation for `X.presheaf`?

Reply:
One possibility to shorten this definition is to define `X.obj_basic_open f := X.presheaf.obj (op $ X.basic_open f)` so that this shortens to  `X.obj_basic_open f ⟶ (Scheme.Spec.obj $ op $ X.presheaf.obj (op U)).obj_basic_open (Spec_Γ_identity.inv.app _ f)`.

Another option is `𝒪 X U := X.presheaf.obj (op U)`, and this becomes
`𝒪 X (X.basic_open f) ⟶ 𝒪 (Scheme.Spec.obj $ op $ 𝒪 X U) (Scheme.basic_open _ $ Spec_Γ_identity.inv.app _ f`. But I don't know if this notation is intuitive enough, or should we go with something like `𝒪_ X U` or `Γ(X, U)`.